File: assets/games/game2/game.py
import json
import os

# 读取地图,不同关卡换json文件的名字！！
with open("map1.json", "r") as f:
    map1 = json.load(f)

class Player:
    def __init__(self, name, start_map="map1_1"):
        self.name = name
        self.map_name = start_map
        self.hp = 3  # 初始生命
        self.pos = self.find_start_pos(start_map)

    def find_start_pos(self, map_name):
        grid = map1[map_name]["grid"]
        for i, row in enumerate(grid):
            for j, val in enumerate(row):
                if val == 3:  # 入口
                    return [i, j]
        raise ValueError(f"{map_name} 没有入口!")

    def display_map(self):
        """在控制台打印地图和玩家位置"""
        os.system('cls' if os.name == 'nt' else 'clear')  # 清屏
        print(f"地图：{self.map_name}   生命：{self.hp}")
        grid = map1[self.map_name]["grid"]
        for i, row in enumerate(grid):
            line = ""
            for j, val in enumerate(row):
                if [i, j] == self.pos:
                    line += " P "
                else:
                    line += f" {val} "
            print(line)
        print("\n方向: up/down/left/right, exit退出")

    def move(self, direction):
        dx, dy = 0, 0
        if direction == "up": dx = -1
        elif direction == "down": dx = 1
        elif direction == "left": dy = -1
        elif direction == "right": dy = 1
        else:
            print("请输入有效方向")
            return False

        x, y = self.pos
        new_x, new_y = x + dx, y + dy
        grid = map1[self.map_name]["grid"]

        # 边界检测
        if not (0 <= new_x < len(grid) and 0 <= new_y < len(grid[0])):
            print("碰到地图边界！")
            return False

        cell = grid[new_x][new_y]

        # 胜利判定
        if cell == 4:  # 出口
            print("🎉 你到达出口，游戏胜利！")
            return "win"

        # 【改动一】切换地图机关：将新位置的坐标传递给 switch_map
        if cell == 2:
            self.switch_map(new_x, new_y)
            return False
        
        # 墙格子判断
        if cell == 1:
            print("撞到了墙上，无法通过！")
            return False # 直接返回，不移动也不掉血

        # 普通可走区域
        self.pos = [new_x, new_y]
        return False

    def switch_map(self,switch_x,switch_y):
        # 切换规则 map1 <-> map2
        new_map = "map1_2" if self.map_name == "map1_1" else "map1_1"
        self.map_name = new_map
        self.pos = [switch_x, switch_y]
        print(f"通过机关切换到 {self.map_name}，新位置 {self.pos}")

        # 检查入口是否在墙上
        grid = map1[new_map]["grid"]
        if grid[self.pos[0]][self.pos[1]] == 1:
            self.hp -= 1
            print(f"入口在墙上！掉一条命，剩余生命：{self.hp}")
            if self.hp <= 0:
                print("💀 生命归零，游戏失败！")
                return "lose"

# 游戏循环
player = Player("玩家1")

while True:
    player.display_map()
    cmd = input("你的操作: ").strip()
    if cmd == "exit":
        break
    result = player.move(cmd)
    if result in ["win", "lose"]:
        break

File: assets/games/game3/game.js
// game.js
document.addEventListener('DOMContentLoaded', () => {
    // --- 全局变量和常量 ---
    const TILE_SIZE = 160; // 必须和CSS中的 --tile-size 一致
    const TILE_CLASS_MAPPING = {
        0: 'floor',
        //1: 'wall',
        2: 'switch',
        3: 'start',
        4: 'exit',
    };

    let mapsData = null; // 用来存储从 map.json 加载的地图数据
    let player = {
        name: "玩家1",
        hp: 3,
        map_name: "map1",
        pos: [0, 0] // 初始位置，稍后会更新
    };
    let isGameOver = false;

    // 获取HTML元素
    const gridContainer = document.getElementById('map-grid');
    const hpDisplay = document.getElementById('hp-display');
    const mapDisplay = document.getElementById('map-display');

    // --- 游戏函数 ---

    // 绘制整个游戏界面
    function drawGame() {
        if (!mapsData) return;

        const grid = mapsData[player.map_name].grid;
        
        // 动态设置网格布局的列数和行数
        gridContainer.style.gridTemplateColumns = `repeat(${grid[0].length}, ${TILE_SIZE}px)`;
        gridContainer.style.gridTemplateRows = `repeat(${grid.length}, ${TILE_SIZE}px)`;
        
        // 清空旧的地图
        gridContainer.innerHTML = '';

        // 遍历地图数据，创建每一个方块div
        grid.forEach((row, i) => {
            row.forEach((tileVal, j) => {
                const tileDiv = document.createElement('div');
                let tileClass = TILE_CLASS_MAPPING[tileVal] || 'floor';
                if (tileVal === 1) {
                    if (player.map_name === 'map1') {
                        tileClass = 'wall-map1';
                    } else { // 假设除了map1之外的地图都用第二种墙壁
                        tileClass = 'wall-map2';
                    }
                }

                tileDiv.classList.add('tile', tileClass);
                
                // 给每个方块加上ID，方便定位玩家
                tileDiv.id = `tile-${i}-${j}`;

                gridContainer.appendChild(tileDiv);
            });
        });

        // 在对应的方块上显示玩家
        const playerTile = document.getElementById(`tile-${player.pos[0]}-${player.pos[1]}`);
        if (playerTile) {
            playerTile.classList.add('player');
        }

        // 更新UI显示
        hpDisplay.textContent = `生命: ${player.hp}`;
        mapDisplay.textContent = `地图: ${player.map_name}`;
    }

    // 处理玩家移动
    function movePlayer(direction) {
        if (isGameOver) return;

        let dx = 0, dy = 0;
        if (direction === "up") dx = -1;
        else if (direction === "down") dx = 1;
        else if (direction === "left") dy = -1;
        else if (direction === "right") dy = 1;

        const [x, y] = player.pos;
        const new_x = x + dx;
        const new_y = y + dy;
        const grid = mapsData[player.map_name].grid;

        if (!(new_x >= 0 && new_x < grid.length && new_y >= 0 && new_y < grid[0].length)) {
            console.log("碰到地图边界！");
            return;
        }

        const cell = grid[new_x][new_y];

        if (cell === 1) {
            console.log("撞到了墙上，无法通过！");
			return;
        }

        if (cell === 4) {
            alert("🎉 你到达出口，游戏胜利！");
            isGameOver = true;
			window.location.href = '../../../game.html?portal-2d-toLoad={"url": "day3.json", "state": 1}';
            return;
        }
        
        if (cell === 2) {
            switchMap(new_x, new_y);
        } else {
            player.pos = [new_x, new_y];
        }

        // --- 新增的被困判断逻辑 ---
        if (isPlayerTrapped()) {
            // 惩罚：直接返回 map1 的起点
            console.log("玩家被困住了! 返回最初的起点。");
            player.map_name = 'map1';                   // 1. 强制切换回 map1
            player.pos = findStartPos('map1');        // 2. 寻找 map1 的入口并设置位置
        }
        // --- 逻辑结束 ---

        // 每次移动后都重绘游戏
        drawGame();

        if (player.hp <= 0) {
            alert("💀 生命归零，游戏失败！");
			window.location.href = '../../../game.html?portal-2d-toLoad={"url": "day3.json", "state": 0}';
            isGameOver = true;
        }
    }

    // 切换地图
    function switchMap(switch_x, switch_y) {
        player.map_name = (player.map_name === "map1") ? "map2" : "map1";
        player.pos = [switch_x, switch_y];

        const grid = mapsData[player.map_name].grid;
        if (grid[player.pos[0]][player.pos[1]] === 1) {
            player.hp -= 1;
            console.log("传送点在墙上！掉一条命");
                        // --- 新增的动画触发逻辑 ---
            hpDisplay.classList.add('damage-flash'); // 立即添加动画类
            
            // 设置一个定时器，在动画播放完毕后移除这个类
            // 动画时长0.5秒，播放2次，总时长1秒 (1000毫秒)
            setTimeout(() => {
                hpDisplay.classList.remove('damage-flash');
            }, 1000);
            // --- 逻辑结束 ---
        }
    }
    
    function isPlayerTrapped() {
        const [x, y] = player.pos;
        const grid = mapsData[player.map_name].grid;

        // 定义上下左右四个相邻格子的坐标
        const neighbors = [
            { r: x - 1, c: y }, // 上
            { r: x + 1, c: y }, // 下
            { r: x, c: y - 1 }, // 左
            { r: x, c: y + 1 }  // 右
        ];

        // 检查所有相邻格子
        for (const neighbor of neighbors) {
            const { r, c } = neighbor;

            // 检查这个邻居坐标是否在地图范围内
            if (r >= 0 && r < grid.length && c >= 0 && c < grid[0].length) {
                // 如果在范围内，检查它是不是墙
                if (grid[r][c] !== 1) {
                    // 只要有一个邻居不是墙，玩家就没有被困住
                    return false;
                }
            }
            // 如果邻居坐标超出了地图边界，我们同样视其为一堵“墙”
        }

        // 如果循环结束，所有邻居都是墙或边界，说明玩家被困住了
        return true;
    }

    // 寻找初始位置
    function findStartPos(map_name) {
        const grid = mapsData[map_name].grid;
        for (let i = 0; i < grid.length; i++) {
            for (let j = 0; j < grid[i].length; j++) {
                if (grid[i][j] === 3) {
                    return [i, j];
                }
            }
        }
        return [0, 0]; // 如果找不到，默认位置
    }

    // --- 游戏启动 ---
    async function initGame() {
        // 异步加载地图文件
        const response = await fetch('第一关map.json');
        mapsData = await response.json();
        
        // 设置玩家初始位置
        player.pos = findStartPos(player.map_name);

        // 首次绘制游戏
        drawGame();

        // 监听键盘事件
// --- 新的键盘事件监听逻辑 ---

        // 帮助函数，用于更新屏幕上虚拟按键的样式
        function updateKeyAppearance(key, isPressed) {
            let keyElementId = null;
            switch(key) {
                case 'ArrowUp': case 'w': keyElementId = 'key-up'; break;
                case 'ArrowDown': case 's': keyElementId = 'key-down'; break;
                case 'ArrowLeft': case 'a': keyElementId = 'key-left'; break;
                case 'ArrowRight': case 'd': keyElementId = 'key-right'; break;
            }

            if (keyElementId) {
                const keyElement = document.getElementById(keyElementId);
                if (isPressed) {
                    keyElement.classList.add('pressed');
                } else {
                    keyElement.classList.remove('pressed');
                }
            }
        }

        // 监听按键按下事件
        window.addEventListener('keydown', (e) => {
            updateKeyAppearance(e.key, true); // 更新按键外观为“按下”

            let direction = null;
            switch(e.key) {
                case 'ArrowUp': case 'w': direction = 'up'; break;
                case 'ArrowDown': case 's': direction = 'down'; break;
                case 'ArrowLeft': case 'a': direction = 'left'; break;
                case 'ArrowRight': case 'd': direction = 'right'; break;
            }
            if (direction) {
                movePlayer(direction); // 移动玩家
            }
        });

        // 监听按键抬起事件
        window.addEventListener('keyup', (e) => {
            updateKeyAppearance(e.key, false); // 恢复按键外观为“抬起”
        });
    }

    initGame();
});